proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1205 @@
+  # was stored.
+  if len(wasm_results) > 1:
+    if direction == "import"
+      wasm_params.insert(0, i32)
+      wasm_results = []
+    else:
+      wasm_results = [i32]
+
+  return wasm_params, wasm_results
+
+def flatten(direction, interface_type):
+  match interface_type
+    f32 => [f32]
+    f64 => [f64]
+    s8 | s16 | s32 | u8 | u16 | u32 => [i32]
+    s64 | u64 => [i64]
+    char => [i32]
+
+    # List are represented by their pointer/length
+    string |
+    list $ty => [i32, i32]
+
+    # flags are represented as a contiguous list of `i32` values
+    flags $fields => {
+      fields = []
+      for _ in 0..num_packed_i32s($fields):
+        fields.push(i32)
+      fields
+    }
+
+    # Records have their fields flattened individually
+    record $fields => $fields.flat_map(|t| flatten(direction, t))
+
+    # variants are represented with a discriminant plus whatever is the
+    # smallest which fits all possible cases.
+    variant $cases => [i32] + flatten_cases(direction, $cases)
+
+    # Handles are represented as a single integer unconditionally. The
+    # un-forgeability is discussed later.
+    handle => [i32]
+
+    # Buffers are represented as three i32 values when passing to an import
+    # and as an integer handle when receiving as part of an export. For more
+    # information on its representation in imports see the comments in `lift`
+    # below.
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        [i32, i32, i32]
+      else:
+        [i32]
+
+# Determines the core wasm types used to represent a variant's cases. This
+# overlays various cases into the same list of types, so the length of types is
+# the length of the longest case, and types are "unified" to a common
+# representation to ensure there's enough bits to hold enough values for all
+# cases.
+def flatten_cases(direction, cases):
+  result = []
+  for case in cases:
+    if case:
+      tys = flatten(direction, case)
+      for i, ty in enumerate(tys):
+        if i < len(result):
+          result[i] = unify(result[i], ty)
+        else:
+          result.push(ty)
+  return result
+
+def unify(a, b):
+  match (a, b)
+    (i64, _) |
+    (_, i64) |
+    (i32, f64) |
+    (f64, i32) => i64,
+
+    (i32, i32) |
+    (i32, f32) |
+    (f31, i32) => i32,
+
+    (f32, f32) => f32,
+    (f64, f64) | (f32, f64) | (f64, i32) => f64,
+
+def num_packed_i32s(fields):
+  return align_to(fields.len(), 32) / 32
+```
+
+The only difference in exports/imports currently is that `push-buffer` and
+`pull-buffer` are represented with a singular `i32` instead of three `i32`
+values. This is because if an export is receiving a buffer it's coming from
+somewhere else which means it's a handle to a buffer and various "intrinsics",
+listed below, will be used to operate on the handle.
+
+## Lifting
+
+The `lift` function here takes four parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
```

Comment:
As an opportunity for simplification for readers here, it seems that using `"result" or "param"` in place of `"export" or "import"` here and adapting the pseudo code would avoid the whole paraphrase, and would be a bit more straightforward. In general, what do you think of using "result" and "param" everywhere as possible values for the `direction` parameter? (It might look a bit weird in the `*_from_memory` functions, though, but wouldn't be too terrible either)